### ISSUE TITLE
Rust vs Python

### ISSUE THREAD

**Issue by VikashKothary** [ankicommunity/anki-sync-server#17]
This sync server is written in Python. Anki is a combination of Python and Rust.

The current plan is to keep this project entirely Python. If this would to change, it should be documented in this issue.

**Comment by VikashKothary** [comment]
So there's now a dependency on Rust code due to the `anki` pypi module. See #26.

**Comment by VikashKothary** [comment]
So I've been giving this one a think as I've seeing an increase in the adoption of Rust in the industry. Here are my thoughts. 

The benefit of a Rust backend would be the improved runtime performance compared to Python. This is useful for some high performance use cases. These might be use cases for time critical or high performance computations. This comes at the tradeoff of being more technical with a higher barrier to entry. 

It also limits the number of platforms that can run this project because Rust code needs to be compiled to the system architecture that it need to run on. This is fine for the original Anki software which is mainly used through Ankiweb.net which is managed by technical developers. However, for a self-hosted version of Anki, this might be limiting to our users. 

The original version of this software was written in pure Python (mainly because the version of Anki it used was written in pure Python). It is my belief that we still to this for the foreseeable future. This is to promote the development and adoption of this project.

However, there is a very clear and present value of including the Rust dependencies that come with the `anki` module. Better performance, reduced development time by reusing existing functions, tightly coupling with the `AnkiWeb` project to ensure long term compatibility. So I feel like there may be a time in the future where we will need to adopt Rust in the project.

Until a time where this project is mature enough to benefit from these features, I believe that the position of the community should be preferential towards pure Python implementations. And all Rust components should be optional. 

The impact of this decision still needs to be analysed but I wanted to document it for posterity and transparency. Have a different take on this subject. Please feel free to share in this issue.

**Comment by AntonOfTheWoods** [comment]
@VikashKothary , while I suppose we can only guess, my belief is that the client code we see is now only used for clients and not used by Ankiweb, at least not a lot of it. One other issue I have always had is with using an embedded, highly non-standard database on the server. 

Both these points call into question the relevance of looking at the upstream code at all. There are definitely some things that *appear* to save time by using upstream but I am not convinced. An API should be a black box, and we really shouldn't care how upstream implements something on the client - we just want the same result for Ankiweb.

Additionally, Rust really is a lot more of a change of thinking than any other language I have looked at. While definitely  increasing, it is still very minor compared to Python, and no one is likely to spend weeks learning just to help out here. Unless you really know what you are doing, even adding a print of a variable is completely non-trivial and utterly non-intuitive. I am definitely too busy for the next several years to learn Rust, for example.

**Comment by VikashKothary** [comment]
@AntonOfTheWoods Oh damn. That is a really good point. I was making the assumption that the conversion to Rust code was to improve the performance in AnkiWeb but you make a very good point.

**Comment by cecini** [comment]
So opt to use rust now.

**Comment by AntonOfTheWoods** [comment]
I agree that rust has advantages. However, from what I can see, there is not an avalanche of devs jumping on to fix bugs or implement new functionality, with either Python or Rust. Making any decisions before having any active devs seems a little like manager-speak :-).

**Comment by VikashKothary** [comment]
@cecini You've made an interesting point.

I understand that hearing the benefits of Rust may make you wonder why we're not getting a jump start on the benefits now. Rust being a compiled language introduces a lot of issues for this projects goals to support a range of self-hosted devices. While Python being interpreted is perfect for that use case. 

For that reason, we plan to use pure Python for all core features. And until all core features have been implemented we will not be looking into the more specialised use cases that would benefit from low-level support as I've mentioned above.

Tbh when we get to the point when we get to the point where we need low-level language support, Rust might not even be the way we go. A lot of JVM languages would have similar benefits but fit better with this project's philosophy of running anywhere. 

It's important to keep in mind (any why I often come back to this ticket), because there is a constant temptation of using Rust code from the original Anki project in this one. While both projects do similar things, they both have different goals and philosophies.

Back to Rust, @AntonOfTheWoods makes a good point, when we decide to talk about the more specialised use cases, we will need to create a call-to-action for Rust developers because there are much rarer. It's important to note that most Rust developers should be able to pick up Python extremely easily but most Python developers will never use Rust.

Does that make sense? What do you think?

**Comment by she3o** [comment]
I stumbled upon this https://github.com/dobefore/anki-sync-server-rs

AND IT COMPILES!!!

**Comment by VikashKothary** [comment]
Yes, that is true. @dobefore has created a Rust version of the sync server and will be sharing it with this community. We should see the repo moving to this organisation very soon :)
